src/app/src/main/java/com/example/bookmark/util/RequestUtil.java

```diff
@@ -14,28 +16,22 @@
  * @author Nayan Prakash
  */
 public class RequestUtil {
-    public static Request retrieveRequestsOnBookByStatus(Book book, Request.Status status, Context context) {
-        Semaphore semaphore = new Semaphore(0);
-        // a final single-element Request array is used to allow assignment within onSuccessListener
-        final Request[] request = {null};
+    public static void retrieveRequestsOnBookByStatus(
+        Book book,
+        Request.Status status,
+        OnSuccessListener<Request> onSuccessListener,
+        OnFailureListener onFailureListener
+    ) {
         StorageServiceProvider.getStorageService().retrieveRequestsByBook(
             book,
             requestList -> {
                 for (Request r: requestList) {
                     if (r.getStatus().equals(status)) {
-                        request[0] = r;
-                        break;
+                        onSuccessListener.onSuccess(r);
```

Comment:
You should return immediately after the first match is found, assuming you only want to get one match. Otherwise the OnSuccessListener might trigger multiple times.
```suggestion
                        onSuccessListener.onSuccess(r);
                        return;
```